include/albatross/src/models/patchwork_gp.hpp

```diff
@@ -0,0 +1,374 @@
+    };
+    const auto C_dd = fit_models.apply(get_train_covariance);
+
+    // C_db holds the covariance between each model and all boundaries.
+    // The actual storage is effectively a map with values which correspond
+    // to the covariance between that model's features and the boundaries.
+    auto C_db_one_group = [&](const auto &key, const auto &fit_model) {
+      const auto group_features =
+          as_group_features(key, fit_model.get_fit().train_features);
+      return this->patchwork_covariance_matrix(group_features,
+                                               boundary_features);
+    };
+    const auto C_db = fit_models.apply(C_db_one_group);
+    const auto C_dd_inv_C_db = block_diag_solve(C_dd, C_db);
+
+    //    S_bb = C_bb - C_db * C_dd^-1 * C_db
+    const Eigen::MatrixXd S_bb =
+        C_bb - block_inner_product(C_db, C_dd_inv_C_db);
+    const auto S_bb_ldlt = S_bb.ldlt();
+
+    auto get_obs_vector = [](const auto &fit_model) {
+      return fit_model.predict(fit_model.get_fit().train_features).mean();
+    };
+
+    const auto ys = fit_models.apply(get_obs_vector);
+    const auto information = patchwork_solver(C_dd, C_db, S_bb_ldlt, ys);
+
+    Eigen::VectorXd C_bb_inv_C_bd_information =
+        Eigen::VectorXd::Zero(C_bb.rows());
+    auto accumulate_C_bb_inv_C_bd_information = [&](const auto &key,
+                                                    const auto &C_bd_i) {
+      C_bb_inv_C_bd_information +=
+          C_bb_ldlt.solve(C_bd_i.transpose() * information.at(key));
+    };
+    C_db.apply(accumulate_C_bb_inv_C_bd_information);
+
+    return ReturnType(*this, PatchworkFitType(fit_models, information,
+                                              boundary_features, C_dd, C_db,
+                                              C_bb_ldlt, S_bb_ldlt));
+  };
+
+  template <typename FeatureType, typename FitModelType, typename GroupKey,
+            typename BoundaryFeatureType>
+  JointDistribution _predict_impl(
+      const std::vector<FeatureType> &features,
+      const Fit<PatchworkGPFit<FitModelType, GroupKey, BoundaryFeatureType>>
+          &patchwork_fit,
+      PredictTypeIdentity<JointDistribution> &&) const {
+
+    if (patchwork_fit.fit_models.size() == 1) {
+      return patchwork_fit.fit_models.values()[0].predict(features).joint();
+    }
+
+    auto predict_grouper = [&](const auto &f) {
+      return patchwork_functions_.nearest_group(
+          patchwork_fit.C_db.keys(), patchwork_functions_.grouper(f));
+    };
+
+    auto group_features = as_group_features(features, predict_grouper);
+
+    const Eigen::MatrixXd C_fb = patchwork_covariance_matrix(
+        group_features, patchwork_fit.boundary_features);
+    const Eigen::MatrixXd C_fb_bb_inv =
+        patchwork_fit.C_bb_ldlt.solve(C_fb.transpose()).transpose();
+
+    auto compute_cross_block_transpose = [&](const auto &key,
+                                             const auto &fit_model) {
```

Comment:
Is there a reason why these lambda functions aren't just `static` functions elsewhere in the file? Personally, I think the readability suffers when these functions are in the middle of other functions rather than moved out of them.

Reply:
Yeah there are a few reasons.  I actually started a slack conversation about this a while back and we settled on the use of lambda functions, even for much simpler cases than this.  The reasons were that 1) There's a bit of computational efficiency that comes out of using lambda over free functions (due to pushing things on the stack) 2) the lambda approach keeps all the functions that are used in the same place their being used. 3) not using either lambdas or free functions would lead to long for loops with comments which is discouraged.

In this particular situation there's another thing going on.  Namely, this lambda function is used with the `apply` method which takes care of a lot of boiler plate around looping over maps.  For example, if I didn't use `apply` I'd have to write something like:
```
std::map<GroupKey, Eigen::MatrixXd> cross_transpose;
for (const auto &pair : fit_models) {
  cross_transpose[pair.first] = cross_transpose(compute_cross_block_transpose(pair.first, pair.second));
}
```
instead of `const auto cross_transpose = fit_models.apply(compute_cross_block_transpose)`.  And if I went the for loop route I probably would have ended up skipping the definition of `compute_cross_block_transpose` and just doing it all in place which leads to a much less readable chunk of code.

I *could* switch this to use free functions, which would look something like defining:
```
template <typename GroupKey, typename FitModelType, typename PatchworkCovarianceFunction>
inline auto compute_cross_block_transpose(const GroupKey &key, const FitModelType &fit_model, const PatchworkCovarianceFunction &patchwork_covariance, const std::map<GroupKey, Eigen::MatrixXd> &C_db) {...};
```
And then I'd have to bind the free function to use it with apply which would end up looking like:
```
auto f = std::bind(compute_cross_block_transpose, _1, _2, patchwork_covariance_matrix, patchwork_fit.C_db);
const auto cross_transpose = patchwork_fit.fit_models.apply(compute_cross_block_transpose);
```
But I don't think that really makes anything clearer, and definitely makes the type deduction more difficult to follow.